docs/rfcs/1011-non-blocking-invite-transport.md

```diff
@@ -0,0 +1,915 @@
+
+In this case, the `invite_greeter_start_attempt` command cancels the active attempts automatically (using a dedicated reason), and joins the new active attempt. This means that the last device calling `start_attempt` always has the priority.
+
+If no cancellation of the active attempt is required, the greeter simply joins it and the command returns the corresponding attempt ID.
+
+Pseudo code:
+
+```python
+invite_greeter_start_attempt(
+    token: InvitationToken,
+)
+->
+  InvitationNotFound | InvitationCompleted | InvitationCancelled
+  GreeterNotAllowed | OK[AttemptID]:
+    # If the greeter has already joined the active attempt:
+    #    -> cancel the active attempt and create a new one
+    [...]
+    # Get the active attempt and call `greeter_join_attempt()`
+    [...]
+    # Return the attempt ID
+    [...]
+```
+
+Schema definition:
+
+```json5
+// TODO
+```
+
+
+#### `invite_claimer_start_attempt` command
+
+This is the new invited command introduced for the claimer to start an attempt and get the corresponding attempt ID.
+
+In order for the claimer to start an attempt, a couple of checks have to be performed:
+- the author exists
+- the author is allowed (i.e. it is part of the greeters)
+
+Note that the state of the invitation doesn't have to be checked since it is already checked by the `/invited` HTTP route.
+
+Otherwise this command has the same logic as `invite_greeter_start_attempt`.
+
+Pseudo code:
+
+```python
+invite_claimer_start_attempt(
+    token: InvitationToken,
+    greeter: UserID,
+)
+->
+  InvitationNotFound | InvitationCompleted | InvitationCancelled
+  GreeterNotFound | GreeterNotAllowed |
+  OK[AttemptID]:
+    # Perform checks
+    [...]
+    # If the claimer has already joined the active attempt:
+    #    -> cancel the active attempt and create a new one
+    [...]
+    # Get the active attempt and call `claimer_join_attempt()`
+    [...]
+    # Return the attempt ID
+    [...]
+```
+
+Schema definition:
+
+```json5
+// TODO
+```
+
+
+#### `invite_greeter_cancel_attempt` command
+
+Two new commands are introduced to cancel an attempt, which registers a cancellation reason for the given attempt. For the greeter, an authenticated command is exposed. After using `invite_greeter_cancel_attempt`, the greeter can start a new attempt by using the `invite_greeter_start_attempt` command.
+
+In order for a greeter to cancel an attempt, a couple of checks have to be performed:
+- the invitation exists
+- the invitation is pending (i.e. not completed or cancelled)
+- the author is allowed (i.e. it is part of the greeters)
+- the attempt exists
+- the attempt is active (i.e. joined and not cancelled)
+
+Then, the active attempt is cancelled with the provided reason and a new active attempt is created. The cancellation reason covers several cases:
+- the user manually cancelled the attempt
+- the hashed nonce didn't match the provided nonce
+- the SAS code communicated to the user was invalid
```

Comment:
So if the user misclicks, they must start over from the beginning ? Could there be a way to just go back to the previous state ?

Reply:
This is up to the client logic, but there has to be a reason defined so that the client is able to communicate this information to the server (i.e "I'm cancelling this attempt because the SAS code didn't match, indicating there might be a MitM in the channel").

In practice, we only display 4 code, so recovering on misclicks might lead to weaker SAS code verification (e.g imagine an impatient user just trying the different code until they find the right one).